bokeh/util/serialization.py

```diff
@@ -54,31 +54,38 @@ def transform_series(obj):
     vals = obj.values
     return transform_array(vals)
 
-def flattened_and_shape(array):
+def transform_array_to_list(array):
+    if np.isnan(array).any() or np.isinf(array).any():
+        transformed = array.astype('object')
+        transformed[np.isnan(array)] = 'NaN'
+        transformed[np.isposinf(array)] = 'Infinity'
+        transformed[np.isneginf(array)] = '-Infinity'
+        return transformed.tolist()
+    return array.tolist()
+
+def serialize_array(array):
+    """Transforms array into one of two serialization formats
+    either a list or a dictionary containing the base64
+    encoded data along with the shape and dtype of the data.
+    """
     if isinstance(array, np.ma.MaskedArray):
         array = array.filled(np.nan)  # Set masked values to nan
     array_samples = np.product(array.shape)
     if (not settings.use_binary_arrays() or
         array_samples < settings.binary_array_cutoff() or
         array.dtype.kind in ('U', 'S', 'O') or
         array.dtype.name == 'int64'):
```

Comment:
So another idea might be to wrap all these conditions up into a predicate function as well:
```
if need_list_seralization(array):
    return transform_array_to_list(array)
```
Apart from being a bit more readable, it would also be good to put the function `need_list_seralization` under unit test, so that the conditions for when list serialization happens can be rigorously maintained